### PR TITLE
Add `UnionToIntersection` type

### PR DESCRIPTION
**Pull request by ifiokjr**
## Description

Creates an intersection type from a union type. This has been useful for me in [library](https://github.com/remirror/remirror/blob/b096ed1dd300d682e28f9a7721ad9debbe1c1eb3/%40remirror/core/src/extension-types.ts#L181-L186) code.

### CODE REVIEW

File: source/union-to-intersection.d.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+type Intersection = UnionToIntersection<Union>;
+//=> {a1(): void; b1(): void; a2(argA: string): void; b2(argB: string): void}
+```
+*/
+export type UnionToIntersection<Union> = (
+  // `extends any` is always going to be the case and is used to convert the
+  // `Union` into a [distributive conditional
+  // type](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html#distributive-conditional-types).
+  Union extends any
```
Comment (sindresorhus) [inline source/union-to-intersection.d.ts]: Can you use `unknown` instead of `any`?

File: source/union-to-intersection.d.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    // The union type is used as the only argument to a function since the union
+    // of function arguments is an intersection.
+    ? (distributedUnion: Union) => void
+    // This won't happen.
+    : never
+    // Infer the `Intersection` type since TypeScript represents the positional
+    // arguments of unions of functions as an intersection of the union.
+  ) extends ((mergedIntersection: infer Intersection) => void)
+    ? Intersection
+    : never;
```
Comment (sindresorhus) [inline source/union-to-intersection.d.ts]: Use tab-indentation. Applies everywhere.

File: source/union-to-intersection.d.ts (hunk 1 above)
Comment (ifiokjr) [reply source/union-to-intersection.d.ts]: Who would have thought 😄 

Nice catch!

File: source/union-to-intersection.d.ts (hunk 2 above)
Comment (ifiokjr) [reply source/union-to-intersection.d.ts]: Fixed 😄

**Comment by ifiokjr** [comment]
I forced push without first checking if you'd made changes yourself.

**Comment by ifiokjr** [comment]
The missing commits are resolved now.

**Comment by sindresorhus** [comment]
👍

**Pull request merged**